rfd/0096-helm-chart-revamp.md

```diff
@@ -0,0 +1,436 @@
+
+#### Chart-specific configuration values
+
+Some values are used to configure the Kubernetes resources deploying Teleport.
+When specified they should apply to both auth and proxy deployments.  Those
+values are:
+
+- `podSecurityPolicy`
+- `labels`
+- `highAvailability`
+- `tls`
+- `image`
+- `enterpriseImage`
+- `affinity`
+- `annotations`
+- `extraArgs`
+- `extraEnv`
+- `extraVolumes`
+- `extraVolumeMounts`
+- `imagePullPolicy`
+- `initContainers`
+- `postStart`
+- `securityContext`
+- `priorityClassName`
+- `tolerations`
+- `probeTimeoutSeconds`
+- `teleportVersionOverride`
+- `resources`
+
+A few values will have to be treated differently:
+
+- `persistence` will only apply to the `auth` deployment
+- `service` will only apply to the `proxy` service
+- `serviceAccount.name` will apply to the `auth`, the proxy service account
+  name should be the auth one suffixed with `-proxy`
+
+Some users will need to set different values for auth and proxy pods, the
+following values should be also available under `auth` and `proxy`. Those
+specific values should take precedence over the ones at the root.
+
+- `labels`
+- `highAvailability` (except the certManager section)
+- `affinity`
+- `annotations`
+- `extraArgs`
+- `extraEnv`
+- `extraVolumes`
+- `extraVolumeMounts`
+- `initContainers`
+- `postStart`
+- `tolerations`
+- `teleportVersionOverride`
+- `resources`
+
+#### CustomResources configuration values
+
+Those values control which Teleport Custom Resources are deployed
+by the chart. Those values can be used only when the operator is
+enabled, to reflect this, they should be nested under the `operator`
+field.
+
+Manually maintaining a 1-1 mapping for every CR field adds little
+value and will create additional maintenance and drift risks.
+Each supported CR spec should be exposed under its own field.
+Specifying the API version in the field name will help managing
+version updates with breaking changes. In extreme cases the chart
+could implement translation from previous versions to the new one,
+but this likely should be the operator's responsibility.
+
+The chart should dynamically set the metadata (name, namespace, labels)
+and pass as-is the spec from the value. For example, a valuefile
+creating two users, one role and setting up the github connector
+would look like:
+
+```yaml
+operator:
+  CRs:
+    userV2:
+      alice:
+        roles:
+          - auditor
+      bob:
+        roles:
+          - team-lead
+    roleV5:
+      team-lead:
+        allow:
+          logins: ["root"]
+          kubernetes_groups: ["system:masters"]
+    githubV2:
+      dev-team:
+        client_id: XXXX
+        client_secret: XXXX
+        display: github-dev-team
+        redirect_url: XXXX
+        teams_to_role:
+          - organization: octocats
+            team: dev
+            roles:
+              - access
+          - organization: octocats
+            team: admin
+            roles:
+              - access
+              - team-lead
```

Comment:
Slight aside: having this exact YAML appear in an example or guide in the docs will be incredibly helpful for anyone like me who learns primarily by reading templated examples :)

Reply:
Yeah, as the chart is becoming even more complex, I think it would be a nice addition to give a couple of example values.yaml to help people quickstart most basic use-cases:

- on gke
- on eks
- on eks with aws-lb-controller
- standalone
- standalone etcd
- deploying operator CRs

We could put them in the test/.lint directory and reference them in the README/guide as quickstart examples.

Reply:
Updated RFD to formalize the need of example values: https://github.com/gravitational/teleport/pull/18274/commits/ec8823cc82ab5a41f8861a8be0e344eab74aea85